design-proposals/klip-5-internal-schemas-for-KSQL-engine.md

```diff
@@ -0,0 +1,191 @@
+
+**Status**: In Discussion |
+
+**Discussion**:
+
+**tl;dr:** *Introduce the internal schemas for KSQL engine to decouple the internal identifier names from the
+external field names so KSQL can handle any field name in the input data.*
+
+## Motivation and background
+
+Currently when users declare the schema for a stream or table, KSQL engine uses the provided schema
+to perform query processing on the stream or table data. This means that the engine uses the same
+names in the schema fields to refer to the schema fields in computations such as expression evaluation
+ or filter evaluation. We generate Java code to evaluate such expressions which imposes certain
+ restrictions on variable naming. However, requiring user schemas to comply with variable naming
+ requirements in Java limits the KSQL ability to process any user data. Although in Avro and Delimited
+ formats KSQL won’t have any issues since the schema field names already are compatible with the current
+ KSQL naming requirements, in handling content with JSON format, KSQL would encounter limitations.
+ In JSON format, field names don’t necessarily conform to KSQL naming requirements and a valid JSON
+ message can have field names that cannot be accepted by KSQL. In addition to the above requirement,
+ the field names in KSQL should also comply with the identifier naming requirements in SQL in order to
+ have the statements parse correctly.
+
+In this KLIP, we propose to decouple the internal schema, that KSQL uses for query processing, from
+ the external schema that user content conforms to. The internal KSQL schema will be fully compliant
+ with KSQL requirements and the engine can use it to process queries with no issues. On the other
+ hand, external schema won’t be restricted with KSQL requirements and users can use any content with
+ any naming convention in KSQL.
+
+Users will be able to use quoted identifiers to declare and access fields that do not conform to
+the KSQL identifier requirements. KSQL will generate an internal schema with a one to one mapping
+from the external schema fields into the internal schema fields. All of the computation will use the
+field names from the internal schema.
+
+As an example, consider we want to declare s stream on a topic with values in JSON format. The following
+is a sample message value in this topic:
+
+```json
+{"@ID": 0, "@NAME": "foo", "@AMOUNT": 0.0}
+```
+
+As you can see, two of the field names start with “@” and the third one includes “.” in the field name.
+ Currently we cannot declare a stream or table on this topic since its content does not conform to KQSL
+ field naming requirements. However, we should be able to use quoted identifiers, which is supported by
+ our parser, to declare a stream or table on this topic and write queries over the declared stream or
+ table. Let’s assume we declare a stream on this topic, named foo. The following would be the DDL
+  statement:
+
+```sql
+CREATE STREAM foo
```

Comment:
I think the parser supports back quotes for quoting too.  We should either disable or support & document.

Reply:
Yes, for now I think we can disable back quote and only support double quotes. We can easily add the support for back quotes if we want to later.